LocalFirst Swarm

joshuef opened the pull request:
As laid out nicely here https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#prioritize-local-work-over-new-work-from-a-remote

we now handle local cmds first, with higher priority, keeping latency for these low 

----------
This pull request primarily involves refactoring of the `SwarmCmd` structure into two separate structures: `LocalSwarmCmd` and `NetworkSwarmCmd` in the `sn_networking/src/driver.rs` and `sn_networking/src/lib.rs` files. This modification is aimed at separating local commands from network commands, improving the organization and readability of the code.

Here are the most important changes:

Refactoring of `SwarmCmd`:

* [`sn_networking/src/driver.rs`](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL16-R16): The `SwarmCmd` structure has been replaced with `LocalSwarmCmd` and `NetworkSwarmCmd` structures. This change is reflected in the import statements, function parameters, and function calls throughout the file. [[1]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL16-R16) [[2]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL457-R463) [[3]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL467-R473) [[4]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL598-R605) [[5]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL618-R631) [[6]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL641-R655) [[7]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bR696-R719) [[8]](diffhunk://#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bL690-R732)

* [`sn_networking/src/event/swarm.rs`](diffhunk://#diff-cb0730cf4bf5e7445bc11bf88d45c13b6f28a706a5d1a7822db1e2dd76b1b2dfL10-R10): The `SwarmCmd` structure has been replaced with `LocalSwarmCmd` in the import statements and function calls. [[1]](diffhunk://#diff-cb0730cf4bf5e7445bc11bf88d45c13b6f28a706a5d1a7822db1e2dd76b1b2dfL10-R10) [[2]](diffhunk://#diff-cb0730cf4bf5e7445bc11bf88d45c13b6f28a706a5d1a7822db1e2dd76b1b2dfL547-R547)

* [`sn_networking/src/lib.rs`](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776R34): The `SwarmCmd` structure has been replaced with `LocalSwarmCmd` and `NetworkSwarmCmd` structures. This change is reflected in the import statements, function parameters, and function calls throughout the file. [[1]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776R34) [[2]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L48-R49) [[3]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L166-R185) [[4]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L204-R214) [[5]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L228-R236) [[6]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L250-R258) [[7]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L261-R269) [[8]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L299-R307) [[9]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L310-R318) [[10]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L490-R498) [[11]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L526-R534) [[12]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L585-R593) [[13]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L594-R608) [[14]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L664-R679) [[15]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L732-R740) [[16]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L743-R757) [[17]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L764-R772) [[18]](diffhunk://#diff-8f8a3f47ddf2bc5366e5a9166b1b25f05b985ef63349b9cf22473822a765e776L777-R785)

These changes enhance the separation of concerns in the codebase, making it easier to manage and understand the different types of commands that can be sent in the network.

joshuef commented:
bors r+

joshuef commented:
bors r+

joshuef commented:
bors r-

joshuef commented:
bors r+

joshuef commented:
bors try

oetyng commented:
bors r+

joshuef committed:
chore(networking): rename SwarmCmd -> NetworkSwarmCmd

Precursor to seperating out local operations vs netwoking IO,
so that we can prioritise the local operaitons.

joshuef committed:
feat(networking): split SwarmCmds into local and network

Handle local cmds with higher priority.

RolandSherwin reviewed (approved):
LGTM!

joshuef committed:
fix(node): fix double spend attempt test

now accumulation may not be in the ms instant. we should be tolerant and accept
any error indicative of the failed tx

joshuef committed:
fix(node): fix double spend attempt test to ensure a spend fails

We don't always expect both to error out with a double spend,
one might pass before acumulation, but both sould never pass

joshuef committed:
test(node): improve DoubleSpendAttempt test with final error check

joshuef committed:
chore(networking): update SwarmCmd logging

joshuef committed:
fix(networking): make TriggerIntervalReplication LocalSwarmCmd

Although it produces network messaging, it's triggered internally in the node
So seems safe.

This fixes some test errors due to data being insufficiently replicated

joshuef committed:
test(networking): increase sleep interval before asserting double spend

joshuef committed:
feat(networking): compare record_types when marking fetches as complete

joshuef committed:
feat(networking): basic retry on send_req OutboundError

joshuef committed:
test(networking): increase wait before verifications

joshuef committed:
ci: disable spend simulation while unstable

joshuef committed:
feat(networking): increase speed of register validation

This should allow more PUT attempts in the same RetryStrategy period

maqi commented on sn_networking/src/driver.rs:
```diff
@@ -695,6 +709,30 @@ impl SwarmDriver {
         loop {
+            // Prioritise any local cmds pending.
+            // https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#prioritize-local-work-over-new-work-from-a-remote
+            match self.local_cmd_receiver.try_recv() {
```
the tokio_select will completes when `any of the branch completes`, which means after this `try_recv of local_cmd` , the next handling of `local_cmd` HAVE TO wait till any of the select branch completed.
which is not ideal for the case when there are bunch of local_cmds need to be handled ?

I'd image the next tokio_select will have to be in the style of `try_recv` as well, to make `local_cmds` really prioritized ?

joshuef replied:
@maqi we have a `continue` just below https://github.com/maidsafe/safe_network/pull/1981/files/01ca2b9111f4ddfb8d4ca165b324d8e7c9d13ade#diff-985b0f4f68bd09d2bcb4c2da67ab65c5c68d38d8aac388691e361a295bc8ae4bR723 

So when we get a `LocalSwarmCmd` successfully, we process it first and go to the next loop iter. Which, if there are local cmds, will be processed again. It's only if we `Err` there (ie, there are no more local cmds waiting) that we go to the `select!` macro.

joshuef replied:
That said, I just reread the `select!` docs and it offers `biased` mode. Which actually I think makes this simpler/cleaner and offers us a way to process locally generated network activity `NetworkSwarmCmds` before incoming `SwarmEvent`s, which should also be beneficial.

joshuef committed:
feat(networking): use tokio::select! biased to gain simpler prioritisation

joshuef committed:
fix(client): verify register when retrieving

By default we should verify, this avoids errors arising from stale data
if it comes form one faster machine

joshuef committed:
chore: clippy fixes for rust 1.80
